File: packages/kits/CorpusKit/Tests/CorpusKitTests/BM25Tests.swift
// BM25Tests.swift

import Testing
import Foundation
import SubstrateTypes
@testable import CorpusKit
import CorpusKitProviders
// ─────────────────────────────────────────────────────────────────
// DO NOT REIMPLEMENT SUBSTRATE MATH.
//
// The substrate publishes conformance-gated, byte-identical
// Swift+Rust implementations of every primitive listed in
// docs/engineering/HARNESS_REFERENCE.md. If you
// need SimHash, Hamming, OR-reduce, Fingerprint256 ops, HammingNN
// top-K, HLC, AuditGate, MatrixDecay, AuditLogFold, Bradley-Terry,
// NMF, FFT, eigenvalue centrality, or any other substrate primitive,
// it's already in SubstrateTypes / SubstrateKernel / SubstrateML.
// CI catches drift four ways. See packages/libs/Substrate{Types,
// Kernel,ML}/AGENTS.md.
// ─────────────────────────────────────────────────────────────────

@Suite("BM25Index")
struct BM25Tests {

    func makeIndex() -> BM25Index {
        BM25Index(tokenizer: DeterministicTokenizer())
    }

    func makeChunk(_ text: String, _ id: UUID = UUID()) -> Chunk {
        Chunk(
            id: id,
            sourceID: "doc",
            startOffset: 0,
            length: text.count,
            text: text,
            hlc: HLC(physicalTime: 1, logicalCount: 0, nodeID: 1)
        )
    }

    /// Tokenise a query with the same tokenizer the test index uses.
    func tokens(_ query: String) -> [String] {
        DeterministicTokenizer().keywordTokens(query)
    }

    @Test func emptyIndexReturnsEmpty() async {
        let idx = makeIndex()
        let results = await idx.topK(10, for: tokens("anything"))
        #expect(results.isEmpty)
    }

    @Test func findsTermInIndexedChunk() async {
        let idx = makeIndex()
        let c1 = makeChunk("the quick brown fox jumps over the lazy dog")
        let c2 = makeChunk("a completely unrelated sentence about cats")
        await idx.index([c1, c2])
        let hits = await idx.topK(5, for: tokens("fox"))
        #expect(!hits.isEmpty)
        #expect(hits.first?.id == c1.id, "fox-bearing chunk should rank first")
    }

    @Test func higherTFRanksHigher() async {
        let idx = makeIndex()
        let c1 = makeChunk("cat cat cat cat cat")
        let c2 = makeChunk("cat and one other thing")
        await idx.index([c1, c2])
        let hits = await idx.topK(5, for: tokens("cat"))
        #expect(hits.first?.id == c1.id, "higher TF should rank first")
    }

    @Test func removeCleansPostings() async {
        let idx = makeIndex()
        let c = makeChunk("ephemeral document content")
        await idx.index([c])
        let count1 = await idx.documentCount()
        #expect(count1 == 1)
        await idx.remove(c.id)
        let count2 = await idx.documentCount()
        #expect(count2 == 0)
        let hits = await idx.topK(5, for: tokens("ephemeral"))
        #expect(hits.isEmpty)
    }
}

File: packages/kits/CorpusKit/rust/tests/bm25_tests.rs
// Tests for BM25Index.

use corpus_kit::{BM25Index, BM25Parameters};
use corpus_kit_providers::DeterministicTokenizer;
use std::sync::Arc;
use uuid::Uuid;

fn make_index() -> BM25Index {
    BM25Index::new(Arc::new(DeterministicTokenizer::new()))
}

/// Tokenise a query with the same tokenizer the test index uses.
fn tokens(idx: &BM25Index, query: &str) -> Vec<String> {
    idx.tokenize_query(query)
}

#[test]
fn empty_index_returns_empty_results() {
    let idx = make_index();
    assert_eq!(idx.document_count(), 0);
    assert!(idx.top_k(10, &tokens(&idx, "anything")).is_empty());
}

#[test]
fn single_document_match() {
    let mut idx = make_index();
    let id = Uuid::new_v4();
    idx.index_documents(vec![(id, "the quick brown fox jumps over the lazy dog")]);
    let toks = tokens(&idx, "quick fox");
    let results = idx.top_k(10, &toks);
    assert_eq!(results.len(), 1);
    assert_eq!(results[0].0, id);
    assert!(results[0].1 > 0.0);
}

#[test]
fn multiple_docs_rank_by_relevance() {
    let mut idx = make_index();
    let a = Uuid::new_v4();
    let b = Uuid::new_v4();
    let c = Uuid::new_v4();
    idx.index_documents(vec![
        (a, "alpha beta gamma delta"),
        (b, "alpha alpha alpha alpha"),
        (c, "epsilon zeta eta theta"),
    ]);
    let toks = tokens(&idx, "alpha");
    let results = idx.top_k(3, &toks);
    assert_eq!(results.len(), 2);
    // Document b has 4 occurrences of "alpha" so it must come first.
    assert_eq!(results[0].0, b);
    assert_eq!(results[1].0, a);
}

#[test]
fn search_with_limit() {
    let mut idx = make_index();
    for i in 0..5 {
        let text = format!("keyword document {}", i);
        idx.index_documents(vec![(Uuid::new_v4(), text.as_str())]);
    }
    let toks = tokens(&idx, "keyword");
    let results = idx.top_k(2, &toks);
    assert_eq!(results.len(), 2);
}

#[test]
fn remove_drops_document_from_results() {
    let mut idx = make_index();
    let id = Uuid::new_v4();
    idx.index_documents(vec![(id, "removable content here")]);
    assert_eq!(idx.document_count(), 1);
    idx.remove(id);
    assert_eq!(idx.document_count(), 0);
    assert!(idx.top_k(10, &tokens(&idx, "removable")).is_empty());
}

#[test]
fn custom_bm25_parameters() {
    let tok = Arc::new(DeterministicTokenizer::new());
    let mut idx = BM25Index::with_parameters(tok, BM25Parameters::new(2.0, 0.5));
    let id = Uuid::new_v4();
    idx.index_documents(vec![(id, "custom parameters test")]);
    let toks = tokens(&idx, "custom");
    let results = idx.top_k(5, &toks);
    assert_eq!(results.len(), 1);
}

#[test]
fn query_with_no_matching_terms_returns_empty() {
    let mut idx = make_index();
    idx.index_documents(vec![(Uuid::new_v4(), "alpha beta gamma")]);
    let toks = tokens(&idx, "zeta");
    let results = idx.top_k(10, &toks);
    assert!(results.is_empty());
}

#[test]
fn empty_query_returns_empty_results() {
    let mut idx = make_index();
    idx.index_documents(vec![(Uuid::new_v4(), "alpha beta")]);
    let toks = tokens(&idx, "");
    let results = idx.top_k(10, &toks);
    assert!(results.is_empty());
}
